### PR TITLE
add cedilla

### PR DESCRIPTION
**Pull request by kevinbarabash**
screen shot 2018-01-31 at 9 58 45 pm

TODO:
- [x] update `makeUnicode.js` to prevent duplicate keys
- [x] update `Unicode` screenshot

### CODE REVIEW

File: metrics/mapping.pl [hunk 1]
```diff
@@ -1189,50 +1197,6 @@
   ],
 };
 
-$map{cmss10} = {
```
Comment (kevinbarabash) [inline metrics/mapping.pl]: I just noticed that this appears twice so I removed the one that didn't match makeFF.

File: src/unicodeMake.js [hunk 2]
```diff
@@ -47,10 +52,14 @@ for (const letter of letters) {
...
-                    `    ${encode(normalized2)}: ${encode(combined2)},`
-                    + `  // ${normalized2} = ${accents[accent].text}`
-                    + `${accents[accent2].text}{${letter}}`);
+                const key = encode(normalized2);
+                if (!keys.has(key)) {
+                    keys.add(key);
+                    console.log(
+                        `    ${encode(normalized2)}: ${encode(combined2)},`
+                        + `  // ${normalized2} = ${accents[accent].text}`
+                        + `${accents[accent2].text}{${letter}}`);
```
Comment (kevinbarabash) [inline src/unicodeMake.js]: The reason why there are duplicates is b/c `\c\'{c}` and `\'\c{c}` map to the same character.  Same goes for the uppercase version.

File: src/functions/accent.js [hunk 3]
```diff
@@ -63,9 +63,11 @@ const htmlBuilder = (group, options) => {
...
 
     // calculate the amount of space between the body and the accent
-    const clearance = Math.min(
-        body.height,
-        options.fontMetrics().xHeight);
+    let clearance = group.value.label === "\\c"
+        ? body.height + body.depth
+        : Math.min(
+            body.height,
+            options.fontMetrics().xHeight);
```
Comment (edemaine) [inline src/functions/accent.js]: Is there a reason that this needs to be special-cased for `\c`?  Can we come up with a general rule?  Ah, I see, there's a fundamental difference according to whether the accent is above or below...  Maybe there should be an `const accentBelow = (group.value.label === "\\c")` which we'll later add more cases to?  Alternatively, `const accentBelow = (accent.depth > 0)` might work...

File: src/functions/accent.js [hunk 4]
```diff
@@ -87,6 +89,9 @@ const htmlBuilder = (group, options) => {
             // shift the accent over to a place we don't want.
             accent.italic = 0;
             width = accent.width;
+            if (group.value.label === "\\c") {
+                clearance += accent.depth;
+            }
```
Comment (edemaine) [inline src/functions/accent.js]: I'm guessing `accent.depth` is zero for the other accents?  Still, might make sense to keep this wrapped in an `if` (possibly `if (accentBelow)`).

File: src/unicodeMake.js (hunk 2 above)
Comment (edemaine) [reply src/unicodeMake.js]: Thanks for fixing this!

I checked that there are still no 3-accent combos that resolve to a single character.

```js
            for (const accent3 of Object.getOwnPropertyNames(accents)) {
                if (accent === accent3 || accent2 === accent3) {
                    continue;
                }
                const combined3 = combined2 + accent3;
                const normalized3 = combined2.normalize('NFC');
                if (normalized3.length === 1) {
                    const key = encode(normalized3);
                    if (!keys.has(key)) {
                        keys.add(key);
                        console.log(
                            `    ${encode(normalized3)}: ${encode(combined3)},`
                            + `  // ${normalized3} = ${accents[accent].text}`
                            + `${accents[accent2].text}{${letter}}`
                            + `${accents[accent3].text}{${letter}}`);
                    }
                }
            }
```

File: src/unicodeMake.js [hunk 5]
```diff
@@ -31,14 +31,19 @@ console.log("export default {");
...
+            const key = encode(normalized);
+            if (!keys.has(key)) {
+                keys.add(key);
+                console.log(
+                    `    ${encode(normalized)}: ${encode(combined)},`
+                    + `  // ${normalized} = ${accents[accent].text}{${letter}}`);
+            }
         }
         for (const accent2 of Object.getOwnPropertyNames(accents)) {
             if (accent === accent2) {
```
Comment (edemaine) [inline src/unicodeMake.js]: I wonder if we should remove this `if (accent === accent2) { continue }` block.  Maybe there is an accent that can be doubled into a single character.  (I checked, and there isn't, but maybe better to have the check in the code.)

File: src/unicodeSymbols.js [hunk 6]
```diff
@@ -35,23 +38,27 @@ export default {
...
     "\u0117": "\u0065\u0307",  // ė = \.{e}
+    "\u0229": "\u0065\u0327",  // ȩ = \c{e}
     "\u1e1f": "\u0066\u0307",  // ḟ = \.{f}
     "\u01f5": "\u0067\u0301",  // ǵ = \'{g}
     "\u1e21": "\u0067\u0304",  // ḡ = \={g}
     "\u011f": "\u0067\u0306",  // ğ = \u{g}
     "\u01e7": "\u0067\u030c",  // ǧ = \v{g}
     "\u011d": "\u0067\u0302",  // ĝ = \^{g}
     "\u0121": "\u0067\u0307",  // ġ = \.{g}
+    "\u0123": "\u0067\u0327",  // ģ = \c{g}
```
Comment (edemaine) [inline src/unicodeSymbols.js]: This one is not going to render correctly (I assume); see @ylemkimon's [recent comment](https://github.com/KaTeX/katex-fonts/pull/26#issuecomment-362148045).  Perhaps we should fix this in another PR, though.  (It'll need another tweak to the font, I suspect.)

**Comment by kevinbarabash** [comment]
@edemaine thanks for the review.  I'll try to make these changes over the weekend.

File: src/functions/accent.js (hunk 3 above)
Comment (kevinbarabash) [reply src/functions/accent.js]: Good call.

File: src/unicodeSymbols.js (hunk 6 above)
Comment (kevinbarabash) [reply src/unicodeSymbols.js]: I'm not sure what the resolution is for `\c{g}`.  Let's handle it separately.

File: src/unicodeMake.js (hunk 5 above)
Comment (kevinbarabash) [reply src/unicodeMake.js]: Let's leave it since it isn't hurting anything.

**Comment by kevinbarabash** [comment]
@edemaine this is ready for review again.

**Comment by edemaine** [comment]
Sorry, missed the request to review this a second time. Can you resolve the conflicts first?

**Comment by kevinbarabash** [comment]
I should've done that after merging that other diff.  Thanks.

File: src/fontMetrics.js [hunk 7]
```diff
@@ -104,7 +104,6 @@ const extraCharacterMap = {
     'Ð': 'D',
     'Þ': 'o',
     'å': 'a',
-    'ç': 'c',
```
Comment (kevinbarabash) [inline src/fontMetrics.js]: I should remove the uppercase one as well.

Comment (kevinbarabash) [reply src/fontMetrics.js]: Done.

**Comment by kevinbarabash** [comment]
@edemaine this is ready for review again.

File: test/unicode-spec.js [hunk 8]
```diff
@@ -64,6 +77,32 @@ describe("unicode", function() {
...
+                        result.message = "'" + actual + "' failed to " +
+                            "build with error: " + e.message;
+                    } else {
+                        result.message = "'" + actual + "' failed " +
+                            "building with unknown error: " + e.message;
+                    }
+                }
+
+                return result;
+            },
```
Comment (edemaine) [inline test/unicode-spec.js]: Sad to see all this code duplication. I guess more motivation to work on #1126! 😄

File: test/unicode-spec.js [hunk 9]
```diff
@@ -72,9 +111,14 @@ describe("unicode", function() {
...
 
+    it("should build Latin-1 inside \\text{}", function() {
+        expect('\\text{ÀÁÂÃÄÅÈÉÊËÌÍÎÏÑÒÓÔÕÖÙÚÛÜÝàáâãäåèéêëìíîïñòóôõöùúûüýÿ' +
+            'ÆÇÐØÞßæçðøþ}').toBuild();
+    });
+
     it("should parse Latin-1 outside \\text{}", function() {
         expect('ÀÁÂÃÄÅÈÉÊËÌÍÎÏÑÒÓÔÕÖÙÚÛÜÝàáâãäåèéêëìíîïñòóôõöùúûüýÿ' +
-            'ÇÐÞçðþ').toParse();
+            'ÐÞðþ').toParse();
```
Comment (edemaine) [inline test/unicode-spec.js]: This is backwards incompatibility: we used to (poorly) support `çÇ` in math mode. I'm not sure what's best here. At the least, removing support would mean a new major version. But we don't have good support to begin with, so maybe that's a good thing.

An alternative would be to define a macro like `\mathcedilla` (by analogy to `\mathring`) for math-mode cedilla and then we could support it in math mode. But I couldn't find any such macro in LaTeX.  We could provide one in `katex.sty` (#793) though...

File: test/unicode-spec.js [hunk 10]
```diff
@@ -72,9 +111,14 @@ describe("unicode", function() {
             'ÆÇÐØÞßæçðøþ}').toParse();
     });
 
+    it("should build Latin-1 inside \\text{}", function() {
+        expect('\\text{ÀÁÂÃÄÅÈÉÊËÌÍÎÏÑÒÓÔÕÖÙÚÛÜÝàáâãäåèéêëìíîïñòóôõöùúûüýÿ' +
+            'ÆÇÐØÞßæçðøþ}').toBuild();
+    });
+
```
Comment (edemaine) [inline test/unicode-spec.js]: This seems to be a duplicate of the above -- perhaps a merge typo?

File: test/unicode-spec.js (hunk 9 above)
Comment (kevinbarabash) [reply test/unicode-spec.js]: We're dropping support for `çÇ` in math mode in #1117.  I rebase after it gets merged.

File: test/unicode-spec.js (hunk 10 above)
Comment (kevinbarabash) [reply test/unicode-spec.js]: This is uses `.toBuild()` whereas the one above uses `.toParse()`.  https://github.com/Khan/KaTeX/issues/1126 should address this.

File: test/unicode-spec.js [hunk 11]
```diff
@@ -81,7 +124,7 @@ describe("unicode", function() {
 
     it("should parse Latin-1 outside \\text{}", function() {
         expect('ÀÁÂÃÄÅÈÉÊËÌÍÎÏÑÒÓÔÕÖÙÚÛÜÝàáâãäåèéêëìíîïñòóôõöùúûüýÿ' +
-            'ÇÐÞçðþ').toParse({unicodeTextInMathMode: true});
+            'ÆÇÐØÞßæçðøþ).toParse({unicodeTextInMathMode: true});
```
Comment (kevinbarabash) [inline test/unicode-spec.js]: bad merge :\

File: test/unicode-spec.js [hunk 12]
```diff
@@ -81,7 +125,7 @@ describe("unicode", function() {
 
     it("should parse Latin-1 outside \\text{}", function() {
         expect('ÀÁÂÃÄÅÈÉÊËÌÍÎÏÑÒÓÔÕÖÙÚÛÜÝàáâãäåèéêëìíîïñòóôõöùúûüýÿ' +
-            'ÇÐÞçðþ').toParse({unicodeTextInMathMode: true});
+            'ÐÞðþ').toParse({unicodeTextInMathMode: true});
```
Comment (kevinbarabash) [inline test/unicode-spec.js]: This test should pass with `ÆÇÐØÞßæçðøþ`.  I'm going to see if I can fix that.

File: src/symbols.js [hunk 13]
```diff
@@ -739,7 +740,8 @@ for (let i = 0; i < letters.length; i++) {
 // but they are not actually in the font, nor are they supported by the
 // Unicode accent mechanism, so they fall back to Times font and look ugly.
 // TODO(edemaine): Fix this.
-export const extraLatin = "ÇÐÞçþ";
+export const extraLatin = "ÐÞþ";
+
```
Comment (edemaine) [inline src/symbols.js]: You'll need to add the cedilla c's to the math symbol table similar to the "ð" below, unless we add a math version of the accent.

Comment (edemaine) [reply src/symbols.js]: Somehow I thought we merged this already, but it looks like the above is still an outstanding issue.  We have two options:
1. Add a `\mathcedilla` or other nonstandard operator for a math version of the cedilla accent, and add support for `ç` etc. in math mode.
2. Add the following lines next to the similar lines for "ð":
```js
 defineSymbol(text, main, mathord, "ç", "ç");
 defineSymbol(text, main, mathord, "Ç", "Ç");
```

The former approach will give the proper glyphs; the latter approach will maintain backward compatibility by forcing çÇ to still fall back to system fonts.

**Comment by kevinbarabash** [comment]
I'm going to start a fresh PR for this.

**Pull request closed without merging**